docs/docs/reference/release-notes/migrating-from-v3-to-v4.md

```diff
@@ -0,0 +1,298 @@
+Now it is changed in favor of interface inheritance:
+
+```diff:title=gatsby-node.js
+exports.createSchemaCustomization = ({ actions }) => {
+  const { createTypes } = actions
+  createTypes(`
+-   interface Foo @nodeInterface
++   interface Foo implements Node
+    {
+      id: ID!
+    }
+  `)
+}
+```
+
+### Setting values on module context outside of `onPluginInit`
+
+Sites and in particular plugins that rely on setting values on module context to access them later in other lifecycles will need to use `onPluginInit`. This is also the case for when you use `onPreInit` or `onPreBootstrap`. The `onPluginInit` API will run in each worker as it is initialized and thus each worker then has the initial plugin state.
+
+Here's an example of a plugin fetching a GraphQL schema at the earliest stage in order to use it in later lifecycles:
+
+```js:title=gatsby-node.js
+const stateCache = {}
+
+const initializePlugin = async (args, pluginOptions) => {
+  const res = await getRemoteGraphQLSchema()
+  const graphqlSdl = await generateSdl(res)
+  const typeMap = await generateTypeMap(res)
+
+  stateCache['sdl'] = graphqlSdl
+  stateCache['typeMap'] = typeMap
+}
+
+// highlight-start
+exports.onPreBootstrap = async (args, pluginOptions) => {
+  await initializePlugin(args, pluginOptions)
+}
+// highlight-end
+
+exports.createResolvers = ({ createResolvers }, pluginOptions) => {
+  const typeMap = stateCache['typeMap']
+
+  createResolvers(generateResolvers(typeMap))
+}
+
+exports.createSchemaCustomization = ({ actions }, pluginOptions) => {
+  const { createTypes } = actions
+
+  const sdl = stateCache['sdl']
+
+  createTypes(sdl)
+}
+```
+
+In order to make this work for Gatsby 4 & Parallel Query Running the logic inside `onPreBootstrap` must be moved to `onPluginInit`:
+
+```js:title=gatsby-node.js
+// Rest of initializePlugin stays the same
+
+exports.onPluginInit = async (args, pluginOptions) => {
+  await initializePlugin(args, pluginOptions)
+}
+
+// Schema APIs stay the same
+```
+
+This also applies to using the `reporter.setErrorMap` function. It now also needs to be run inside `onPluginInit` instead of in `onPreInit`.
+
+```js
+const ERROR_MAP = {
+  10000: {
+    text: context => context.sourceMessage,
+    level: "ERROR",
+    category: "SYSTEM",
+  },
+}
+
+// highlight-start
+exports.onPluginInit = ({ reporter }) => {
+  reporter.setErrorMap(ERROR_MAP)
+}
+// highlight-end
+
+const getDataFromAPI = async ({ reporter }) => {
+  let data
+  try {
+    const res = await requestAPI()
+    data = res
+  } catch (error) {
+    reporter.panic({
+      id: "10000",
+      context: {
+        sourceMessage: error.message,
+      },
+    })
+  }
+
+  return data
+}
+```
+
+### Removal of obsolete flags
+
+The config flags for `QUERY_ON_DEMAND` and `PRESERVE_WEBPACK_CACHE` have been removed since the features are enabled by default. Thus you can't explicitly opt-out of these anymore.
+
+### Node mutation & creation in custom resolvers
+
+TODO
```

Comment:
I'd be grateful about some content here from you @vladar - can be rough/draft, I'll polish it.

Will mark you in other spots as well :)